src/format/opamFile.ml

```diff
@@ -551,26 +551,97 @@ module Environment = LineFile(struct
     let internal = "environment"
     let atomic = true
 
-    type t = env_update list
+    type t = resolved_gf env_update list
 
     let empty = []
 
     let pp =
-      (OpamFormat.lines_set ~empty:[] ~add:OpamStd.List.cons ~fold:List.fold_right @@
-       Pp.identity ^+
-       Pp.of_pair "env_update_op"
-         (OpamLexer.FullPos.env_update_op, OpamPrinter.env_update_op_kind) ^+
-       Pp.identity ^+
-       Pp.opt Pp.singleton)
+      let host_or_target : (string, host_or_target) Pp.t =
+        Pp.of_pair "host-or-target"
+          ((function
+              | "host" -> Host
+              | "target" -> Target
+              | "target-quoted" -> Target_quoted
+              | "host-quoted" -> Host_quoted
+              | _ -> Pp.unexpected ()),
+           OpamTypesBase.string_of_host_or_target)
+      in
+      let separator : (string, separator) Pp.t =
+        Pp.of_pair "separator"
+          ((function
+              | ":" -> SCol
+              | ";" -> SSemiCol
+              | _ -> Pp.unexpected ()),
+           (fun sep -> String.make 1 (char_of_separator sep)))
+      in
+      let env: (string, env_update_op_kind) Pp.t =
+        (Pp.of_pair "env_update_op"
+           (OpamLexer.FullPos.env_update_op, OpamPrinter.env_update_op_kind))
+      in
+      let comment_and_sepformat =
+        (separator ^+ host_or_target ^+ (Pp.last -| Pp.identity))
+        -| (Pp.of_pair "comment-sep-ht"
+              ((fun (sep, (ht, comment)) ->
+                  (Some comment, Some (sep, ht))),
+               (function | Some comment, Some (sep, ht) -> sep, (ht, comment)
+                         | _, _ -> failwith "")))
```

Comment:
there should probably be a proper message here, or `assert false` if it should not happend.

Reply:
for the record: it can't be assert false here as we use `Pp.fallback` that errors on `Assertion_failure`. Added a message to failwith.

Reply:
maybe it need simply `Pp.unexpected ()`